### PR TITLE
Only shuffle the tx builders if both flags are enabled

### LINKED ISSUE #931: Setting TransactionBuilder.ShuffleOutputs to false doesn't prevent output shuffling
## Steps to reproduce:

Create a `TransactionBuilder`, set its `ShuffleOutputs` flag to false, add some outputs using `Send` then call `BuildTransaction`.

## Expected behaviour:

The outputs appear in the order that they were added using `Send`.

## Actual behaviour:

The outputs appear in a random order.

---

Note that it's possible to fix this by also setting `TransactionBuilder.RandomShuffle` to `null`. But that will disable all shuffling (including input shuffling) and it also makes it unnecessary to set `ShuffleOutputs` at all. So as it stands the `ShuffleOutputs` flag is just broken.

From a brief look at the code, it seems that the bug is caused by `TransactionBuilder.BuildTransaction` calling `DoShuffle` at the very start of the function, regardless of whether the `ShuffleOutputs` flag is set. It later shuffles the outputs again depending on whether `ShuffleOutputs` is set, but by that point the outputs have already been shuffled.

### PR DESCRIPTION
**Pull request by aarani**
Fixes #931

### CODE REVIEW

**Comment by NicolasDorier** [comment]
Can you instead change the `DoShuffle` method to check if shuffling is authorized and output shuffling as well?

**Comment by knocte** [comment]
@aarani CI is broken now, pls check when you. have time

**Comment by knocte** [comment]
@aarani also in your last push you removed the removal of the DoStuff() method at the beginning, so the bug would still be there.

File: NBitcoin/TransactionBuilder.cs [hunk 1]
```diff
@@ -1153,7 +1153,10 @@ public TransactionBuilder Shuffle()
 
 		private void DoShuffle()
 		{
-			if (ShuffleRandom != null)
+			TransactionBuildingContext ctx = new TransactionBuildingContext(this);
+			if (ShuffleRandom != null &&
+				ShuffleInputs && ctx.CanShuffleInputs &&
```
Comment (NicolasDorier) [inline NBitcoin/TransactionBuilder.cs:1158]: remove this, this DoShuffle only shuffle outputs.

**Comment by NicolasDorier** [comment]
The test not passing is me that broke the test. Fixing.

File: NBitcoin/TransactionBuilder.cs (hunk 1 above)
Comment (knocte) [reply NBitcoin/TransactionBuilder.cs:1158]: then let's rename DoShuffle to DoShuffleOutputs?

**Comment by knocte** [comment]
> @aarani also in your last push you removed the removal of the DoStuff() method at the beginning,

Disregard this bit, I misread the last diff; the bug would be fixed by the last diff but maybe it's more readable to put the two shuffles closer to each other.

File: NBitcoin/TransactionBuilder.cs [hunk 2]
```diff
@@ -1153,7 +1153,9 @@ public TransactionBuilder Shuffle()
 
 		private void DoShuffle()
 		{
-			if (ShuffleRandom != null)
+			TransactionBuildingContext ctx = new TransactionBuildingContext(this);
```
Comment (NicolasDorier) [inline NBitcoin/TransactionBuilder.cs:1156]: remove this, this is uneeded.

Comment (knocte) [reply NBitcoin/TransactionBuilder.cs:1156]: @NicolasDorier how can he check ctx.CanShuffleOutputs then? should pass ctx as param?

**Comment by NicolasDorier** [comment]
made new version

**Pull request merged**